Remove controller from put API header

riya-17 opened the pull request:
This PR deals with removing controller from PUT API Header.

dbutenhof reviewed (commented):
> > + black --check .
would reformat lib/pbench/server/cache_manager.py

webbnh reviewed (changes requested):
I really think that we should be removing the Server-side validation of the controller name.  And, along those lines, I think we should start removing any special treatment of it (e.g., we shouldn't be passing it to the `CacheManager`, as far as I can tell).  I think the revisions to the `Tarball` methods are nicely done, but I recommend going at least one step further.  (And, I have a couple of nits.)

dbutenhof commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -398,18 +403,6 @@ def _put(self, args: ApiParams, request: Request, context: ApiContext) -> Respon
             # NOTE: we're setting the Dataset "created" timestamp here, but it
             # won't be committed to the database until the "advance" operation
             # at the end.
-            try:
-                metadata = tarball.get_metadata()
-                dataset.created = UtcTimeHelper.from_string(
-                    metadata["pbench"]["date"]
-                ).utc_time
-                Metadata.create(dataset=dataset, key=Metadata.METALOG, value=metadata)
-            except Exception as exc:
-                raise CleanupTime(
-                    HTTPStatus.BAD_REQUEST,
-                    f"Tarball {dataset.name!r} is invalid or missing required metadata.log: {exc}",
-                )
-
```
The alternative, which wouldn't really simplify anything but would possibly make the cache manager interface look a tiny bit prettier (I don't believe it's reasonable to care yet), would be to rely on the cache manager `create` itself to unpack `metadata.log` from the tarball. I'd considered that approach, and decided that wasn't a good idea for this pass since `PUT` is already unpacking it.

We are very definitely not ready to remove `controller` from lower levels: it's the basis of the file system structure, and changing this will break the indexer, which we do not want to address in this PR.

webbnh replied:
I'm very much in favor of designing and implementing the _interfaces_ that we want to use, even if they don't immediately yield the best implementation.

If `CacheManager.create()` needs to come up with a "real" value for the controller (at least, in the interim), that's OK with me (although, I suspect that using a constant value would suffice), but I don't think it should be provided as a parameter (especially, since the caller then has to go through the same motions that `create()` would have in order to produce it).

That is, if we're going to implement throw-away code, lets put it exactly where it is needed and not in that function's caller.  (Yes, it might mean duplicating the access to the metadata (or it might not...), but when the controller is no longer needed, we won't have to disturb the calling code when we rip it out.)

dbutenhof replied:
Riya ran into a chicken-and-egg issue, and addressed it cleanly with minimal interface changes. I like that, and in general would like to encourage that rather than taking a superficial change like dropping an HTTP header as an excuse to radically refactor a loosely related class and possibly rototill the historical file system hierarchy.

We'll be doing all that: this isn't the time. The cache manager interfaces will *definitely* be changing, substantially, in order to fulfill its mission as we move into dynamic tracked unpacking. Making an out-of-scope change now won't reduce the churn later: it just increases the churn now.

webbnh replied:
Understood, and I don't disagree...which is why I suggested removing the controller value entirely from `_put()`, rather than taking the extra step to produce it so that we don't have to change the call to `CacheManager.create()`:  given that with this change `_put()` will no longer have direct access to the controller value (i.e., from the HTTP headers), it makes as much sense as anything to relegate the problem/change of producing it to the `CacheManager.create()` code as it does to burden the `_put()` code with it.  (In particular, the result is _less_ change to the `_put()` code, which is the stuff that we're going to "keep".)  And, then `create()` can produce it any way it wants (e.g., it can probably just use a constant rather than trying to come up with the real value).

dbutenhof commented on lib/pbench/server/cache_manager.py:
```diff
@@ -296,8 +298,9 @@ def get_metadata(self) -> JSONOBJECT:
         Returns:
             A JSON representation of the dataset `metadata.log`
         """
+        name = tarball_path.name[0:-len(".tar.xz")]
```
`Dataset.stem` was built for that purpose.

npalaska commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -348,6 +337,22 @@ def _put(self, args: ApiParams, request: Request, context: ApiContext) -> Respon
                     )
 
+            try:
+                metadata = Tarball.get_metadata(tar_full_path)
+                controller = metadata["run"]["controller"]
```
Should this be in a try-except block of its own? That way, if a key is missing, we can propagate an appropriate response back to the user instead of the generic response on line 350.

webbnh replied:
Let's resist the urge to have too many `try` blocks!  🥴  If we need to catch it specially, we should be able to do so by adding a `KeyError` clause to the existing `try` block.  However, @npalaska, why do you think that the existing handling at line 350 is insufficient?  I _think_ it will explicitly report that the `"controller"` key is missing if it is omitted from the metadata.

npalaska replied:
Yeah, I guess that's sufficient.

npalaska commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -348,6 +337,22 @@ def _put(self, args: ApiParams, request: Request, context: ApiContext) -> Respon
+                    f"Tarball {dataset.name!r} is invalid or missing required metadata.log: {exc}",
+                )
+
+            if validate_hostname(controller) != 0:
+                raise CleanupTime(HTTPStatus.BAD_REQUEST, "Invalid 'controller' header")
```
Also, it's not a header anymore!

npalaska reviewed (commented):
Looks good so far

dbutenhof reviewed (approved):
Nice!

portante commented on lib/pbench/server/cache_manager.py:
```diff
@@ -882,6 +890,11 @@ def create(self, controller_name: str, tarfile: Path) -> Tarball:
+            metadata = Tarball.get_metadata(tarfile)
+            controller_name = metadata["run"]["controller"]
+        except Exception as exc:
+            raise KeyError(f"Missing required metadata: {exc}")
```
Why are we turning all errors into KeyError?

portante replied:
Shouldn't this be something like:

```
try:
    metadata = Tarball.get_metadata(tarfile)
    controller_name = metadata["run"]["controller"]
except KeyError as exc:
    raise MetadataError("Missing required metadata") from exc
```

webbnh reviewed (changes requested):
Like Dave, I think these changes are very nice!  However, I think there are a few bits that need some more polishing.

Like Peter, I think the handling of a missing controller metadata key isn't quite right (although, I disagree with him on the details 🙂); I think the lack of symmetry in the access to the `Tarball.metadata` attribute should be reconciled; and, there are a couple of issues with the unit tests.  (And, there's a typographical nit.  😉)

riya-17 commented on lib/pbench/test/unit/server/test_requests.py:
```diff
@@ -59,25 +58,28 @@ class FakeTarball:
             def delete(self):
                 TestUpload.tarball_deleted = self.name
 
             def get_metadata(self):
-                return {"pbench": {"date": "2002-05-16T00:00:00"}}
+                self.metadata = {"pbench": {"date": "2002-05-16T00:00:00"}}
 
```
Whole Cache manager was a fake. So, according to the way it was organized it was behaving correctly because I did not pass the metadata and assigned it directly.

riya-17 commented on lib/pbench/test/unit/server/test_requests.py:
```diff
@@ -107,19 +109,6 @@ def test_malformed_authorization_header(
         assert response.status_code == HTTPStatus.UNAUTHORIZED
         assert not self.cachemanager_created
 
-    def test_missing_controller_header_upload(
-        self, client, caplog, server_config, pbench_token
-    ):
-        expected_message = "Missing required 'controller' header"
-        response = client.put(
-            self.gen_uri(server_config),
-            headers={"Authorization": "Bearer " + pbench_token},
-        )
-        assert response.status_code == HTTPStatus.BAD_REQUEST
-        assert response.json.get("message") == expected_message
-        self.verify_logs(caplog)
-        assert not self.cachemanager_created
-
```
We are moving away from controller, upload API doesn't have any reference to controller and we will be moving forward with the same approach on the Agent.

dbutenhof replied:
True, but there's a similar replacement problem: what if `metadata.log` doesn't have `pbench.run.controller`?? You could test *that* behavior here instead.

riya-17 replied:
but we are not check that here.. whole metadata.log file computation is happening on cache_manager now.. I can add an extra test there.

dbutenhof replied:
Sure; that'd probably be more appropriate given the new structure.

riya-17 commented on lib/pbench/test/unit/server/test_requests.py:
```diff
@@ -189,24 +174,6 @@ def test_bad_length_header_upload(
         self.verify_logs(caplog)
         assert not self.cachemanager_created
 
-    def test_bad_controller_upload(
-        self, client, caplog, server_config, setup_ctrl, pbench_token
-    ):
-        expected_message = "Invalid 'controller' header"
-        response = client.put(
-            self.gen_uri(server_config),
-            headers={
-                "Authorization": "Bearer " + pbench_token,
-                "controller": "not_a_hostname",
-                "Content-MD5": "ANYMD5",
-                "Content-Length": "STRING",
-            },
-        )
-        assert response.status_code == HTTPStatus.BAD_REQUEST
-        assert response.json.get("message") == expected_message
-        self.verify_logs(caplog)
-        assert not self.cachemanager_created
-
```
We are moving away from controller, upload API doesn't have any reference to controller and we will be moving forward with the same approach on the Agent. I am not sure if we sould go forward with this.

dbutenhof commented on lib/pbench/server/api/resources/upload_api.py:
```diff
@@ -63,10 +63,10 @@ def __str__(self) -> str:
-    creates a pair of files (tarball and MD5 file) within the designated
+    Upload a dataset from an agent. This API accepts a tarball and MD5
+    value from a client. After validation, it creates a new Dataset DB
+    row describing the dataset, along with some metadata, and it creates
+    a pair of files (tarball and MD5 file) within the designated
     controller directory under the configured ARCHIVE file tree.
```
> "it creates a pair of files (tarball and MD5 file) within the designated controller directory under the configured ARCHIVE file tree"

Hmm. Technically, it doesn't, and this text outlived its usefulness (and accuracy) a long time ago; that's a function of the cache manager, and shouldn't really be described here. Upload creates temporary files and hands them off to the cache manager "to do what it does". While it's not a big deal now, ideally when the cache manager stuffs the tarball into S3 we shouldn't need to change comments *here* to remain accurate.

dbutenhof commented on lib/pbench/server/cache_manager.py:
```diff
@@ -893,6 +898,7 @@ def create(self, controller_name: str, tarfile: Path) -> Tarball:
             controller = Controller.create(controller_name, self.options, self.logger)
             self.controllers[controller_name] = controller
         tarball = controller.create_tarball(tarfile)
+        tarball.metadata = metadata
```
You could probably get rid of the local `tarball` variable here without annoying `black` too badly ...

portante replied:
How?

dbutenhof replied:
Yeah, never mind.

dbutenhof commented on lib/pbench/server/cache_manager.py:
```diff
@@ -280,13 +281,12 @@ def extract(self, path: str) -> str:
 
-    def get_metadata(self) -> JSONOBJECT:
+    @staticmethod
+    def get_metadata(tarball_path: Path) -> JSONOBJECT:
```
The way this is currently structured, this could probably be `_get_metadata` to discourage external use.

portante commented on lib/pbench/server/cache_manager.py:
```diff
@@ -280,13 +281,12 @@ def extract(self, path: str) -> str:
-            return (
-                tarfile.open(self.tarball_path, "r:*").extractfile(path).read().decode()
-            )
+            return tarfile.open(tarball_path, "r:*").extractfile(path).read().decode()
```
So by using `fs-version-002` we won't break existing `fs-version-001` controller named based hierarchies.

dbutenhof replied:
I'd been avoiding even thinking about departing that radically in the context of *this* PR.

But now that I think about it ... it might not be that bad, since the indexer now relies on the cache manager to look up a `Tarball` by resource ID and looks for the `unpacked` directory there. So does the inventory API; so it's possible that the disruption would be minimal. (All in all I'd still prefer to defer that to a separate PR, if nothing else because I'd like to get my #3194 on top of this.)

And regardless of this, I really don't want to defer capturing `dataset.metalog` and creation timestamp until we unpack. Once we graduate to a real cache manager this'll all be available dynamically as soon as we turn over control from `PUT` to cache manager. In the meantime, this seemed like a reasonable compromise. Of course if we'd rather use `tar` than `tarfile`, we could just do `tar -xfO {tarball} {tarball}/metadata.log` and capture the `stdout` stream.

portante replied:
I'd like to remove all unpacking of the tar ball in the upload API.

The "created" field is not important, since the expiration is set based on the upload time.

If the "created" field is not there (yet) the UI does not have to display it.

That reduces the lifting being done during an already expensive API.

Then the unpack operation can add all that metadata to the dataset and it will show up in the UI.

Once we have a message passing architecture, then we can hide all these steps such that the data set can be in a "processing" mode in the background if the UI catches it in that state.

So I'd like to see a PR remove the unpacking for metadata behavior in the upload API, another PR remove the cache manager use of controller and move to an MD5 hierarchy, and then have this change land.

Landing this now expands the use of the metadata in the upload API, which I don't want to have to back out later.

dbutenhof replied:
1. I'd agree that the `created` column in `Dataset` is no longer critical since I started capturing the entire `metadata.log`; I didn't have that originally and I'd wanted to retain that information. I'd basically tried to avoid random DB schema changes, partly because they tend to ripple through the code and tests. I wouldn't object to removing the column. If I did that, I'd probably move `metalog` from a `Metadata` key inline into `Dataset` as a `JSON` column. I might do `tarball-path` as well, since that'll be our S3 URI handle eventually. (I might rename it, too.)
2. However, having metadata not present for some arbitrary time after upload is inconsistent and awkward. I really don't like that idea, and I think it significantly reduces the value of metadata in general.
3. Removing the controller hierarchy in the filesystem makes a lot of sense. I'd been hoping we'd be able to remove filesystem dependencies with cache manager before we played with that; but we may actually already be there, between `Sync` and `CacheManager` APIs. But the way Riya coded it it's just another access to the information that we've already been capturing for a year now; and I really, really, dislike the idea of dropping the capture of `metadata.log`. We can argue separately about destroying metadata, but I don't see any sense in trying to make this PR suddenly dependent on unrelated changes in that area.

dbutenhof replied:
Pete's always talked about the performance costs of `tarfile`; and it's not that I ever doubted that, but I've always been curious about quantifying the difference.

I wrote a little script to measure and compare average costs of directly extracting `metadata.log` via `tarfile.open().extract()` vs `subprocess.run('tar xf0')`. This  run points at the functional test tarballs (which are small, to be sure), measuring and averaging 500 iterations of each mechanism for each tarball.

As you can see, the actual times are tenths or hundredths of a second, and `tar` generally shows around 50% advantage here. Interesting.

__Anyway, this generates some data, but leaves the essential question of how critical this overhead is right now and how to weigh the tradeoffs. I kinda think that having the metadata available from upload is probably worth the 1/10 second it takes. And having extracted the metadata for `dataset.metalog`, using it again for the controller is "free". If the alternative is taking time to re-implement the file system hierarchy maybe getting the controller is an acceptable tradeoff now?__

```
Starting fio_rw_2018.02.01T22.40.57 (2.2 MB)
	 average tarfile extract time: 0.0736548
	 average tar extract time: 0.0331896
Starting linpack_mock_2020.02.28T19.10.55 (20.8 kB)
	 average tarfile extract time: 0.010313
	 average tar extract time: 0.00819257
Starting pbench-user-benchmark_example-vmstat_2018.10.24T14.38.18 (31.4 kB)
	 average tarfile extract time: 0.0167387
	 average tar extract time: 0.0106804
Starting trafficgen_basic-forwarding-example_tg:trex-profile_pf:forwarding_test.json_ml:5_tt:bs__2019-08-27T14:58:38 (1.0 MB)
	 average tarfile extract time: 0.268283
	 average tar extract time: 0.135689
Starting uperf_rhel8.1_4.18.0-107.el8_snap4_25gb_virt_2019.06.21T01.28.57 (8.4 MB)
	 average tarfile extract time: 0.159724
	 average tar extract time: 0.0774569
Tar wins 0.053 to 0.106 by 0.0527: 49.84%
```

Interesting trivia: on some smaller tarfiles (e.g. the ones on intlab-002 under the `ff00cfa30cd2` controller), the `tarfile` package actually wins by about 18% (a margin of 0.00026 seconds) but the largest file easily available (gandalf's fio run at 21.4 Mb) shows a 56% advantage to `tar`. Here, using `tarfile` adds about 0.17 seconds to `PUT` (switching to `tar` would cut that to 0.07).

I could easily run this on our 0.69 server, and it'd be interesting to see how this escalates on "big" datasets; but I seem to have lost my keys.

portante replied:
I don't agree with this path of using `tarfile` or of unpacking the tarball on the Upload API.  But since others approve and do not seem concerned about it, I'll resolve my comment and withdraw my review.

portante reviewed (changes requested):
We really need to stop using `tarfile` objects in our code, especially on the upload API path.

The upload API should not have to read the contents of the tar ball to do anything with it.  We can have any metadata pulled out of the tar ball and added to the dataset by the "unpack" operation.

And the cache manager can use the proposed storage hierarchy that does not require controller name and is efficient.

portante commented on lib/pbench/server/cache_manager.py:
```diff
@@ -882,6 +882,14 @@ def create(self, controller_name: str, tarfile: Path) -> Tarball:
+        try:
+            metadata = Tarball._get_metadata(tarfile)
+            controller_name = metadata["run"]["controller"]
+        except Exception as exc:
```
I think you have to catch at least MetadataError, and what ever happens when a `metadata.log` file is not found.

webbnh reviewed (dismissed):
I have no problems with the changes as currently proposed.

I think Peter's point that we should be reordering and streamlining (or better compartmentalizing) our sequence of operations is very good (I concur that it would be better if upload didn't have to crack open the tarball), however, what is proposed here seems to work and (per Dave's research) the performance load (at least in the short term) of reading the tarball isn't onerous, so I would suggest that we should defer changes of that magnitude until after our initial Server deployment (unless there is some overriding reason why we cannot or should not do so).

dbutenhof commented on lib/pbench/server/cache_manager.py:
```diff
@@ -882,6 +882,16 @@ def create(self, controller_name: str, tarfile: Path) -> Tarball:
         """
+        try:
+            metadata = Tarball._get_metadata(tarfile)
+            controller_name = metadata["run"]["controller"]
```
I think a `KeyError` here should be caught explicitly rather than falling to the `except Exception`: `extractfile` can raise a `KeyError` because `{name}/metadata.log` doesn't exist (which should really be handled in `Tarball.extract`), *and* we don't know that `metadata` has a `"run"` or that the `run` dict has a `"controller"`. `_get_metadata` should also recognize that we might get `tarfile.TarError` exceptions (it's not really detailed, although it suggests that it usually just returns `None`).

dbutenhof commented on lib/pbench/test/unit/server/test_cache_manager.py:
```diff
@@ -120,50 +125,104 @@ def test_clean_empties(self, server_config, make_logger):
                 assert not (r / c).exists()
 
-    def test_create_bad(self, selinux_disabled, server_config, make_logger, tarball):
+    def test_metadata(
```
The `tarfile` `extractfile` method will apparently raise `KeyError` if the metadata.log target doesn't exist, and return `None` on unspecified other errors. (It's not entirely clear from the documentation whether it might raise `TarError`?) Those cases ought to be tested, and I don't think they are?

webbnh commented on lib/pbench/server/cache_manager.py:
```diff
@@ -885,6 +885,8 @@ def create(self, tarfile: Path) -> Tarball:
+        except MetadataError as exc:
+            raise MetadataError(tarfile, exc)
         except Exception as exc:
             raise MetadataError(tarfile, exc)
```
I don't see any point to having two separate `except` clauses which perform exactly the same action.

We _could_ combine them together:
```
        except (Exception, MetadataError) as exc:
            raise MetadataError(tarfile, exc)
```
But that would be silly, given that any instance of `MetadataError` would match `Exception`, so there is no need to list it as well.

As long as I'm pointing out silly stuff, it looks to me like line 889 is raising another instance of the exception which it just caught.  All that that does is truncate the traceback and obscure the original source of the exception.  If we want to keep the `except` clause at line 888 (e.g., to keep it separate from the subsequent one), then its action should be a bare `raise`, to reraise the original `MetadataError` and keep the information intact.

webbnh reviewed (changes requested):
I don't think that the addition of the `except` clause helps anything -- I think it was better the way it was.  (Now, that said, Dave is probably right about the exception handling and testing being a little weak, but I'm not feeling particularly impelled to have that addressed here...so I'll defer to him and Peter on that.)

webbnh commented on lib/pbench/test/unit/server/test_cache_manager.py:
```diff
@@ -148,6 +154,19 @@ def fake_metadata_controller(tar_path):
         # being there should result in a MetadataError
+        tar_name = source_tarball.name.removesuffix(".tar.xz")
+
+        expected_metaerror = f"A problem occurred processing metadata.log from {source_tarball!s}: \"\'filename {tar_name}/metadata.log not found\'\""
```
This looks odd to me:  do we really need to escape the _single_-quotes when they are inside double-quotes?

webbnh commented on lib/pbench/test/unit/server/test_cache_manager.py:
```diff
@@ -148,6 +154,19 @@ def fake_metadata_controller(tar_path):
+        tar_name = source_tarball.name.removesuffix(".tar.xz")
+
+        expected_metaerror = f"A problem occurred processing metadata.log from {source_tarball!s}: \"\'filename {tar_name}/metadata.log not found\'\""
+        monkeypatch.setattr(Tarball, "extract", fake_extract)
+        with pytest.raises(Exception) as exc:
+            cm.create(source_tarball)
+        assert str(exc.value) == expected_metaerror
```
Should we be either catching something more specific than `Exception` or verifying that `exc` was an instance of some particular exception class?  (Yes, we confirm that it has the right message, which is important for humans, but knowing that it has the right exception instance would be potentially important for calling code.)

Ditto a line 167.

riya-17 replied:
Aren't we going to catch that specific exception through `Exception`, So, shouldn't we try to test the real scenario here?

webbnh replied:
The unit test is supposed to ensure that our CUT behaves as expected.  What do we expect `create()` to do when its behavior is driven by `fake_extract()`?  I doubt it's actually going to raise an `Exception` -- rather, it will raise a subclass of that...so, which subclass do we expect it to raise?  (Put another way, what is the "interface contract" of `create()` in this situation?  Surely, it is more specific than just "raises an exception"?...)  That is what we should be checking for at line 161 and/or in addition to 163.

riya-17 replied:
According to my `fake_extract` we are raising `KeyError`. After `KeyError` is raised with the message from `fake_extract` it is caught by `create()`. `create()` catces the ``KeyError` in `Exception` which is then again raised through `MetadataError`. Raised `MetadataError` includes the message that `KeyError` has included in it and that is caught in pytest.raises(Exception). Either I can include a specific catch phrase for the errors that can be raised, like KeyError or TarError but unless we don't have those, they will be caught in  `Exception` block of create() which is again raised as `MetadataError`.

webbnh replied:
The unit test code here is testing `create()`, so it should be checking the behavior of that function (under the various conditions that we expect it might face).  We expect `create()` to raise `MetadataError`, so the unit test should ensure that it receives that exception object (specifically).  In addition, we can check the message associated with it, if we think that that is important (which, given that that is what you've coded, I think we do think that 🙂).

That said, looking at the body of `Tarball.extract()`, the only exception that it will raise is `MetadataError`, so `fake_extract()` is not a good representation of that function.

Therefore, I recommend two changes:  `fake_extract()` should be changed to raise `MetadataError` instead of `KeyError`, and the `pytest.raises()` call should specify `MetadataError` instead of `Exception`.

If  you agree with that recommendation, then the `fake_extract_tarfile()` needs some consideration:  we should really make the same change to that, for the same reason, in which case it will become identical to `fake_extract()`, in which case it is redundant, and we should just remove it and the code which uses it.

----

Now, this prompts a look at the other mocks which this test is employing, which all replace `Tarball._get_metadata()`.  Since `CacheManager.create()` calls `Tarball._get_metadata()` which in turn calls `Tarball.extract()`, the fact that you're mocking different levels of this call tree at different times calls into question what your definition of "unit" is, here.  That is, does the unit encompass `CacheManager` and everything it calls (in which case, you should really be mocking `tarfile.open()` instead of `Tarball.extract()`) or is it limited to the `CacheManager.create()` method itself (in which case you should be mocking `Tarball._get_metadata()` in every case and control should never reach `Tarball.extract()`)?

And, as long as I'm asking awkward questions, why is a `CacheManager` method calling an _internal_ method of the `Tarball` class?  (I.e., why does the name of `_get_metadata()` begin with an underscore if it is referenced outside the `Tarball` class?)

----

As long as we're talking about unit testing, I encourage you to look at the Cobertura coverage report for your PR (e.g., for [`cache_manager.py`](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3193/22/cobertura/server/cache_manager_py/)).  You've modified `Tarball.extract()`, but there is no unit test coverage for it.  You've modified `Tarball._get_metadata()`, but it's only half covered.  And, in `CacheManager.create()`, there's coverage for only one of the two `except` clauses that you added.

We can argue that `Tarball.extract()` is too simple to bother with unit testing for it, but the counter-argument is that that makes it very easy to test....  🙂  How we get full coverage of `CacheManager.create()` will depend on what you determine to be the "unit", here.

dbutenhof replied:
In terms of `Tarball.extract`, I could definitely see a unit test case with a mocked  `tarfile.extractfile` raising `KeyError`, which I believe is what should happen when you try to extract a non-existent file from the tar; the unit test would be proving that it (or actually any exception from `tarfile`) is reported *outside* as a `MetadataError`.

But as Webb says, when mocking `Tarball.extract`... we always expect `MetadataError` from that. So while I like your `KeyError` and `TarError` cases, Webb's correct that these need to be tested at a lower level by mocking `tarfile` directly instead of `Tarball.extract`.

webbnh replied:
> while I like your `KeyError` and `TarError` cases, Webb's correct that these need to be tested at a lower level by mocking `tarfile` directly instead of `Tarball.extract`.

Or...we could define the "units" to be `CacheManager.create()` and `Tarball.extract()`, respectively, and test them separately.  With that approach, we would need (say) two test scenarios for `Tarball.extract()` (proving that it handles each of `KeyError` and `TarError` correctly), and then only a single test scenario for `CacheManager.create()` (proving that it handles `MetadataError` appropriately).

(My instinct is that, because we went to the trouble of defining separate classes for `CacheManager` and `Tarball`, we should be testing them separately:  we should try to keep the unit testing decoupled for all the same reasons that we try to keep the individual classes decoupled.)

riya-17 replied:
> Therefore, I recommend two changes: `fake_extract()` should be changed to raise `MetadataError` instead of `KeyError`, and the `pytest.raises()` call should specify `MetadataError` instead of `Exception`.

I have fixed this error and removed the Redundant `fake_extract_tarfile()` Error.

webbnh reviewed (approved):
Looks OK to me, although, unfortunately, Black disagrees.

Also, I have a couple of pointed questions.  😕

webbnh reviewed (approved):
👍

dbutenhof reviewed (approved):
Applying `pytest.raises` to the most specific exception class in the unit tests would be really good; given Webb's open conversation on the topic, I'm going to approve anyway since it's suboptimal but arguably not "wrong". 😏

npalaska reviewed (dismissed):
Looks okay, the only thing that remains to see is whether the upload gets affected during opening the large tarfile.

webbnh commented:
> the only thing that remains to see is whether the upload gets affected during opening the large tarfile.

Actually, I think there is general agreement that we would rather not open the tar file during the upload operation (although that leads to the question of how and when does the server create the metadata associated with it).  So, _for this PR_, we'll crack the tarball during the upload, and in a subsequent PR we will iterate and improve the operation.

The [remaining question](https://github.com/distributed-system-analysis/pbench/pull/3193#discussion_r1091096056) is a detail of how to test the cache manager `create()` operation.

That and Riya needs to resolve a conflict between her branch and the current `main`.

npalaska commented:
> (although that leads to the question of how and when does the server create the metadata associated with it)

We would eventually do the unpack on demand so I guess that's when we associate all the metadata to it (when the user actually wants to see the contents).

webbnh commented:
> > (although that leads to the question of how and when does the server create the metadata associated with it)
> 
> We would eventually do the unpack on demand so I guess that's when we associate all the metadata to it (when the user actually wants to see the contents).

But, how would a user know which result to demand if s/he cannot query based on metadata?  🥴

dbutenhof commented:
> > We would eventually do the unpack on demand so I guess that's when we associate all the metadata to it (when the user actually wants to see the contents).
> 
> But, how would a user know which result to demand if s/he cannot query based on metadata? woozy_face

Yeah, well that's why I think we need the metadata from upload. The only real question is whether there's a sufficient performance impact that we need to make compromises on "correctness". I need to get access to "big" datasets to verify the overhead, and there's no guarantee it won't become a problem but for right now I don't think that 1/10 second is a big deal.

When we drop the unpacker in favor of dynamic cache management, the cache manager will still have to start out by cracking the tarball at least to build a TOC, and pulling `metadata.log` at the same time (and this should be early) makes sense.

How to deal with tarballs that don't have `metadata.log` is a different issue, but one option is that we simply ignore it if we can't find it.

None of this should be tied to this PR, however. (Which now needs a rebase before it can be approved...)

webbnh commented:
> The only real question is whether there's a sufficient performance impact that we need to make compromises on "correctness".

No, there's the question of timing: how long should the user have to wait before being able to make a particular query?  If we don't crack the tarball on upload, then, immediately after upload, the only queries which will hit will be the dataset ID, ownership, and similar stuff which is independent of the metadata log file.  Once the Server gets around to processing the tarball, _then_ the other queries will start hitting.  The question is, is this sort of delay a problem or is it just the way it is/will-be?  Presumably, the user will be able to ask if the just-uploaded dataset is ready to be queried...so perhaps it's not a problem to have no metadata available until that answer changes to "yes".  And, that might engage with the no-`metadata.log` case -- the "is it ready" query can return a "no...not now...not ever..." response.

> I need to get access to "big" datasets to verify the overhead, and there's no guarantee it won't become a problem but for right now I don't think that 1/10 second is a big deal.

I think we have one from Dave Valin around somewhere....  Personally, given modern hardware and OS, I don't think it will be a problem -- if we can survive the upload process, I think that the cost of peeking into the tarball will be lost in the noise -- but, I think Peter's point is nonetheless well-taken, and if we start accepting large volumes of data we're going to want to streamline the process, and not cracking the tarball early would be an excellent step.

> None of this should be tied to this PR, however. (Which now needs a rebase before it can be approved...)

Agreed!

webbnh reviewed (dismissed):
I'm not sure what changed in this latest update (other than "something in `test_requests.py`"), because, @riya-17, you modified the existing commit instead of pushing up a new one.  But, I don't have any objections to anything here except for my [previous comment](https://github.com/distributed-system-analysis/pbench/pull/3193#discussion_r1091096056).

dbutenhof reviewed (dismissed):
Nuances of unit testing the cache manager interface contract aside, I think this is looking pretty good and I'd like to get it in.

webbnh commented:
> Nuances of unit testing the cache manager interface contract aside, I think this is looking pretty good

Yeah, but I don't think it has changed in the past week.

> I'd like to get it in.

Are you nudging me or @riya-17??  😆

riya-17 commented:
> Therefore, I recommend two changes: fake_extract() should be changed to raise MetadataError instead of KeyError, and the pytest.raises() call should specify MetadataError instead of Exception.

I have fixed this error and removed the Redundant fake_extract_tarfile() Error.

> In terms of Tarball.extract, I could definitely see a unit test case with a mocked tarfile.extractfile raising KeyError

I have covered this in my next commit and will update that.

riya-17 commented:
@dbutenhof @webbnh I have fixed tests in the latest commit with `tarfile` fakes. I didn't change the ones with `_get_metadata` as it is a wide change. I will try to get back on that in another PR.

dbutenhof commented:
> > Nuances of unit testing the cache manager interface contract aside, I think this is looking pretty good
> 
> Yeah, but I don't think it has changed in the past week.
> 
> > I'd like to get it in.
> 
> Are you nudging me or @riya-17?? laughing

s/or/and ???

webbnh reviewed (approved):
👍
